src/workerd/io/container-client.c++

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2025 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "container-client.h"
+
+#include <kj/array.h>
+#include <kj/debug.h>
+
+namespace workerd::io {
+
+ContainerStreamSharedState::ContainerStreamSharedState() {}
+
+void ContainerStreamSharedState::enqueueMessage(::rust::Slice<const uint8_t> message) {
+  auto lockedQueue = messageQueue.lockExclusive();
+  for (auto& byte : message) {
+    lockedQueue->push(byte);
+  }
+
+  auto lockedWaiter = readWaiter.lockExclusive();
+  KJ_IF_SOME(fulfiller, *lockedWaiter) {
+    fulfiller->fulfill();
+    *lockedWaiter = kj::none;
+  }
```

Comment:
Should both locks actually be held simultaneously? Or should each of these be in their own blocks?